[FSDP] Fix exec order validation for diff ignored modules across ranks

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #79584 [FSDP] Default to not checking exec order
* #79557 [FSDP] Extend ignored modules test to not pass to root
* **#79533 [FSDP] Fix exec order validation for diff ignored modules across ranks**

The execution order validation implementation did not account for ignored modules nor differing ignored modules across ranks. It assumed that all parameters in the root FSDP module were `FlatParameter`s. This PR removes the assumption.

It is still critical that the `FlatParameter`s follow the same order within `fsdp_root.parameters()`, but this should be ensured if the model definition is the same across ranks since the model is recursively-wrapped following an order-stable module traversal.

If we want to support arbitrary `FlatParameter` order within `fsdp_root.parameters()`, then we need more logic. We need something like calling `dist.broadcast_object_list()` to broadcast a data structure that gives rank 0's `FlatParameter` order in `fsdp_root.parameters()`, where `FlatParameter`s are identified by a `str` key constructed from the unflattened names that comprise it. We need such a key to uniquely identify `FlatParameter`s and determine if any two represent the same logical entity.
- FYI: This unique keying by unflattened parameter names is added in my non-recursive wrapping refactor.
- To have differing `FlatParameter` order within `fsdp_root.parameters()`, we need the child `nn.Module`s wrapped in FSDP to be registered in different orders in the original model definition. E.g.:
```
Rank 0:
FSDP(Model(
    FSDP(Layer1()),
    FSDP(Layer2()),
))
Rank 1:
FSDP(Model(
    FSDP(Layer2()),
    FSDP(Layer1()),
))
```

---
Example:
```
[Rank 0]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
            (3): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Note the additional `IgnoredModule` designated by the `---->`.
```
[Rank 1]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
---->       (3): IgnoredModule()
            (4): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Because rank 1 has an additional `IgnoredModule` and `layer3` is wrapped in FSDP, rank 1's root module's `enumerate(parameters())` will assign `layer3`'s `FlatParameter` an index that is one higher than that of rank 0. In the existing execution order validation, this led to an incorrect association of `FlatParameter`s to indices. 

---

Differential Revision: [D37143127](https://our.internmc.facebook.com/intern/diff/D37143127)

Commit:
[FSDP] Fix exec order validation for diff ignored modules across ranks

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37143127).

Commit:
Update on "[FSDP] Fix exec order validation for diff ignored modules across ranks"


Differential Revision: [D37143127](https://our.internmc.facebook.com/intern/diff/D37143127)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37143127).

Commit:
Update on "[FSDP] Fix exec order validation for diff ignored modules across ranks"


Differential Revision: [D37143127](https://our.internmc.facebook.com/intern/diff/D37143127)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37143127).

rohan-varma commented on test/distributed/fsdp/test_fsdp_ignored_modules.py:
```diff
@@ -123,12 +149,31 @@ def test_ignored_modules_invalid(self):
+        model_ignored_modules = [
+            m for m in model.modules() if isinstance(m, IgnoredModule)
+        ]
+        wrapped_model = FSDP(model, ignored_modules=model_ignored_modules)
```
should we also test where ignored modules are passed at non-root?

rohan-varma reviewed (approved):
LGTM, let's test and see if it fixes the issue.

Commit:
Update on "[FSDP] Fix exec order validation for diff ignored modules across ranks"


The execution order validation implementation did not account for ignored modules nor differing ignored modules across ranks. It assumed that all parameters in the root FSDP module were `FlatParameter`s. This PR removes the assumption.

It is still critical that the `FlatParameter`s follow the same order within `fsdp_root.parameters()`, but this should be ensured if the model definition is the same across ranks since the model is recursively-wrapped following an order-stable module traversal.

If we want to support arbitrary `FlatParameter` order within `fsdp_root.parameters()`, then we need more logic. We need something like calling `dist.broadcast_object_list()` to broadcast a data structure that gives rank 0's `FlatParameter` order in `fsdp_root.parameters()`, where `FlatParameter`s are identified by a `str` key constructed from the unflattened names that comprise it. We need such a key to uniquely identify `FlatParameter`s and determine if any two represent the same logical entity.
- FYI: This unique keying by unflattened parameter names is added in my non-recursive wrapping refactor.
- To have differing `FlatParameter` order within `fsdp_root.parameters()`, we need the child `nn.Module`s wrapped in FSDP to be registered in different orders in the original model definition. E.g.:
```
Rank 0:
FSDP(Model(
    FSDP(Layer1()),
    FSDP(Layer2()),
))
Rank 1:
FSDP(Model(
    FSDP(Layer2()),
    FSDP(Layer1()),
))
```

---
Example:
```
[Rank 0]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
            (3): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Note the additional `IgnoredModule` designated by the `---->`.
```
[Rank 1]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
---->       (3): IgnoredModule()
            (4): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Because rank 1 has an additional `IgnoredModule` and `layer3` is wrapped in FSDP, rank 1's root module's `enumerate(parameters())` will assign `layer3`'s `FlatParameter` an index that is one higher than that of rank 0. In the existing execution order validation, this led to an incorrect association of `FlatParameter`s to indices. 

---

Differential Revision: [D37143127](https://our.internmc.facebook.com/intern/diff/D37143127)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37143127).

Commit:
Update on "[FSDP] Fix exec order validation for diff ignored modules across ranks"


The execution order validation implementation did not account for ignored modules nor differing ignored modules across ranks. It assumed that all parameters in the root FSDP module were `FlatParameter`s. This PR removes the assumption.

It is still critical that the `FlatParameter`s follow the same order within `fsdp_root.parameters()`, but this should be ensured if the model definition is the same across ranks since the model is recursively-wrapped following an order-stable module traversal.

If we want to support arbitrary `FlatParameter` order within `fsdp_root.parameters()`, then we need more logic. We need something like calling `dist.broadcast_object_list()` to broadcast a data structure that gives rank 0's `FlatParameter` order in `fsdp_root.parameters()`, where `FlatParameter`s are identified by a `str` key constructed from the unflattened names that comprise it. We need such a key to uniquely identify `FlatParameter`s and determine if any two represent the same logical entity.
- FYI: This unique keying by unflattened parameter names is added in my non-recursive wrapping refactor.
- To have differing `FlatParameter` order within `fsdp_root.parameters()`, we need the child `nn.Module`s wrapped in FSDP to be registered in different orders in the original model definition. E.g.:
```
Rank 0:
FSDP(Model(
    FSDP(Layer1()),
    FSDP(Layer2()),
))
Rank 1:
FSDP(Model(
    FSDP(Layer2()),
    FSDP(Layer1()),
))
```

---
Example:
```
[Rank 0]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
            (3): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Note the additional `IgnoredModule` designated by the `---->`.
```
[Rank 1]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
---->       (3): IgnoredModule()
            (4): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Because rank 1 has an additional `IgnoredModule` and `layer3` is wrapped in FSDP, rank 1's root module's `enumerate(parameters())` will assign `layer3`'s `FlatParameter` an index that is one higher than that of rank 0. In the existing execution order validation, this led to an incorrect association of `FlatParameter`s to indices. 

---

Differential Revision: [D37143127](https://our.internmc.facebook.com/intern/diff/D37143127)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37143127).

Commit:
Update on "[FSDP] Fix exec order validation for diff ignored modules across ranks"


The execution order validation implementation did not account for ignored modules nor differing ignored modules across ranks. It assumed that all parameters in the root FSDP module were `FlatParameter`s. This PR removes the assumption.

It is still critical that the `FlatParameter`s follow the same order within `fsdp_root.parameters()`, but this should be ensured if the model definition is the same across ranks since the model is recursively-wrapped following an order-stable module traversal.

If we want to support arbitrary `FlatParameter` order within `fsdp_root.parameters()`, then we need more logic. We need something like calling `dist.broadcast_object_list()` to broadcast a data structure that gives rank 0's `FlatParameter` order in `fsdp_root.parameters()`, where `FlatParameter`s are identified by a `str` key constructed from the unflattened names that comprise it. We need such a key to uniquely identify `FlatParameter`s and determine if any two represent the same logical entity.
- FYI: This unique keying by unflattened parameter names is added in my non-recursive wrapping refactor.
- To have differing `FlatParameter` order within `fsdp_root.parameters()`, we need the child `nn.Module`s wrapped in FSDP to be registered in different orders in the original model definition. E.g.:
```
Rank 0:
FSDP(Model(
    FSDP(Layer1()),
    FSDP(Layer2()),
))
Rank 1:
FSDP(Model(
    FSDP(Layer2()),
    FSDP(Layer1()),
))
```

---
Example:
```
[Rank 0]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
            (3): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Note the additional `IgnoredModule` designated by the `---->`.
```
[Rank 1]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
---->       (3): IgnoredModule()
            (4): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Because rank 1 has an additional `IgnoredModule` and `layer3` is wrapped in FSDP, rank 1's root module's `enumerate(parameters())` will assign `layer3`'s `FlatParameter` an index that is one higher than that of rank 0. In the existing execution order validation, this led to an incorrect association of `FlatParameter`s to indices. 

---

Differential Revision: [D37143127](https://our.internmc.facebook.com/intern/diff/D37143127)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37143127).

Commit:
Update on "[FSDP] Fix exec order validation for diff ignored modules across ranks"


The execution order validation implementation did not account for ignored modules nor differing ignored modules across ranks. It assumed that all parameters in the root FSDP module were `FlatParameter`s. This PR removes the assumption.

It is still critical that the `FlatParameter`s follow the same order within `fsdp_root.parameters()`, but this should be ensured if the model definition is the same across ranks since the model is recursively-wrapped following an order-stable module traversal.

If we want to support arbitrary `FlatParameter` order within `fsdp_root.parameters()`, then we need more logic. We need something like calling `dist.broadcast_object_list()` to broadcast a data structure that gives rank 0's `FlatParameter` order in `fsdp_root.parameters()`, where `FlatParameter`s are identified by a `str` key constructed from the unflattened names that comprise it. We need such a key to uniquely identify `FlatParameter`s and determine if any two represent the same logical entity.
- FYI: This unique keying by unflattened parameter names is added in my non-recursive wrapping refactor.
- To have differing `FlatParameter` order within `fsdp_root.parameters()`, we need the child `nn.Module`s wrapped in FSDP to be registered in different orders in the original model definition. E.g.:
```
Rank 0:
FSDP(Model(
    FSDP(Layer1()),
    FSDP(Layer2()),
))
Rank 1:
FSDP(Model(
    FSDP(Layer2()),
    FSDP(Layer1()),
))
```

---
Example:
```
[Rank 0]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
            (3): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Note the additional `IgnoredModule` designated by the `---->`.
```
[Rank 1]
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FlattenParamsWrapper(
    (_fpw_module): ModelWithIgnoredModules(
      (layer0): Linear(in_features=3, out_features=5, bias=True)
      (layer1): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Sequential(
            (0): Linear(in_features=5, out_features=4, bias=True)
            (1): Linear(in_features=4, out_features=4, bias=True)
            (2): IgnoredModule()
---->       (3): IgnoredModule()
            (4): Linear(in_features=4, out_features=4, bias=True)
          )
        )
      )
      (layer2): Linear(in_features=4, out_features=2, bias=True)
      (layer3): FullyShardedDataParallel(
        (_fsdp_wrapped_module): FlattenParamsWrapper(
          (_fpw_module): Linear(in_features=2, out_features=2, bias=True)
        )
      )
      (relu): ReLU()
    )
  )
)
```
Because rank 1 has an additional `IgnoredModule` and `layer3` is wrapped in FSDP, rank 1's root module's `enumerate(parameters())` will assign `layer3`'s `FlatParameter` an index that is one higher than that of rank 0. In the existing execution order validation, this led to an incorrect association of `FlatParameter`s to indices. 

---

Differential Revision: [D37143127](https://our.internmc.facebook.com/intern/diff/D37143127)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37143127).

awgu commented:
@pytorchbot merge
